docs/developer-guide/ceps/proposed/cep-004-module-structure.rst

```diff
@@ -0,0 +1,79 @@
+.. _cep-003:
+
+************************************************************
+CEP 4 - Changing the module structure of the ctapipe package
+************************************************************
+
+* Status: draft
+* Discussion: NA
+* Date accepted: NA
+* Last revised: 2024-01-08
+* Author: Maximilian Linhoff
+* Created: 2024-01-08
+
+
+Abstract
+========
+
+``ctapipe``'s module structure has organically grown over its development
+in the last couple of years and has some suboptimal properties.
+
+* Large inter-dependencies between modules, sometimes even circular. E.g. importing ``ctapipe.calib`` will import ``ctapipe.image`` because the image extraction code lives in ``ctapipe.image``.
+* Deep hierarchies, resulting in issues with references in the sphinx documentation due to the same class being documentated at more than two locations.
+
+A better structure would resolve these issues and would also make it easier to
+assign maintainers / code owners to specific submodules.
+
+
+Proposed new structure
+======================
+
+In the new structure, we want to use a flatter hierarchy of smaller modules,
+that will resolve the above mentioned issues.
+
+We expect no disadvantages other than:
+* Breaking compatibility of imports once
+* More namespaces for developers to remember
+
+The second point should be offset by the more logical structure of the namespaces.
+
+.. code::
+
+   ctapipe
+        .atmosphere
+        .camera_calibration
+        .config
+        .coordinates
+        .core
+        .data_model
+        .data_volume_reduction
+        .gain_selection
+        .image_cleaning
+        .image_features
+        .instrument
+        .muon
+        .processors
+        .reconstruction
+        .tools
+        .visualization
+
+
+Looking at it from the current structure, we'd have the following changes:
+
+- ``.image`` is split into ``.image_cleaning`` and ``.image_features``
+- ``.calib`` is split into ``.camera_calibration``, ``.data_volume_reduction``, ``.gain_selection``
+- ``.image.extractor`` is moved to ``.camera_calibration``
```

Comment:
To be more clear:
* camera calibration should contain things that relate to:
  *  DC to PE conversion
  * flat-fielding
  * pedestals
  * noisy pixel detection
  * pointing correction? → maybe in its own module, as this is not camera, but structure related. 
* waveform processing: generation of images from waveforms
   * ImageExtractors 
   * related code for peak detection and pulse characterization in general: time over threshold,FWHM,  risetime, falltime, etc
   * DataVolumeReducers

Reply:
> pointing correction? → maybe in its own module, as this is not camera, but structure related.

Definitely it's own module, there won't be any code overlap, so it would again lead to a situation where people only wanting to import the pointing correction module would import all the low-level calibrations resulting in large import times.

Reply:
> DataVolumeReducers

I added a module specifically for that.

Reply:
Yeah, I noticed that and updated the comment...